File: Java/POJ/2386/Main.java
import java.util.Scanner;

public class Main {
	int n, m;
	char[][] map;
	static final char CHECK = 'x';
	void run() {
		Scanner scan = new Scanner(System.in);
		n = scan.nextInt();
		m = scan.nextInt();
		map = new char[n][];
		for (int i = 0; i < n; i++)
			map[i] = scan.next().toCharArray();
		int cnt = 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (map[i][j] != 'W') continue;
				cnt++;
				solve(i, j);
			}
		}
		System.out.println(cnt);
	}
	
	void solve(int y, int x) {
		if (y < 0 || x < 0 || y >= n || x >= m) return;
		if (map[y][x] != 'W') return;
		map[y][x] = CHECK;
		int[] dy = { -1, -1, -1,  0, 0,  1, 1, 1 };
		int[] dx = { -1,  0,  1, -1, 1, -1, 0, 1 };
		for (int i = 0; i < 8; i++) {
			solve(y + dy[i], x + dx[i]);
		}
	}
	public static void main(String[] args) {
		new Main().run();
	}
}

File: Go/POJ/2386/Main.go
package main

import (
	"fmt"
)

var n, m int
var field [][]int

func main() {
	fmt.Scan(&n)
	fmt.Scan(&m)
	field = make([][]int, n)
	for i := 0; i < n; i++ {
		field[i] = make([]int, m)
		var str string
		fmt.Scan(&str)
		for j := 0; j < m; j++ {
			if string(str[j]) == "W" {
				field[i][j] = 1;
			}
		}
	}
	cnt := 0
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if field[i][j] == 1 {
				cnt++
				Solve(i, j);
			}
		}
	}
	fmt.Println(cnt)
}

func Solve(y, x int) {
	if y < 0 || x < 0 || y >= n || x >= m {
		return
	}
	if field[y][x] != 1 {
		return
	}
	field[y][x] = 0
	dy := []int { -1, -1, -1,  0, 0,  1, 1, 1 }
	dx := []int { -1,  0,  1, -1, 1, -1, 0, 1 }
	for i := 0; i < 8; i++ {
		Solve(y + dy[i], x + dx[i])
	}
}
